src/v/cloud_storage/inventory/ntp_hashes.cc

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cloud_storage/inventory/ntp_hashes.h"
+
+#include "cloud_storage/logger.h"
+#include "container/fragmented_vector.h"
+#include "hashing/xx.h"
+
+#include <seastar/core/fstream.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/util/file.hh>
+
+namespace cloud_storage::inventory {
+ntp_path_hashes::ntp_path_hashes(
+  model::ntp ntp, std::filesystem::path hash_store_path)
+  : _ntp{std::move(ntp)}
+  , _hash_store_path{std::move(hash_store_path)}
+  , _ntp_hashes_path{_hash_store_path / std::string_view{_ntp.path()}}
+  , _rtc{_as}
+  , _ctxlog{cst_log, _rtc, _ntp.path()} {}
+
+ss::future<bool> ntp_path_hashes::load_hashes() {
+    if (co_await ss::file_exists(_ntp_hashes_path.string())) {
+        co_await ss::util::with_file_input_stream(
+          _ntp_hashes_path,
+          [this](auto& stream) { return load_hashes(stream); });
+        _loaded = true;
+        vlog(
+          _ctxlog.trace,
+          "loaded {} hashes from disk, possible collisions {}",
+          _path_hashes.size(),
+          _possible_collisions.size());
+        co_return true;
+    }
+
+    vlog(_ctxlog.warn, "hash file {} not found", _ntp_hashes_path.string());
+    co_return false;
+}
+
+ss::future<> ntp_path_hashes::load_hashes(ss::input_stream<char>& stream) {
+    while (!stream.eof()) {
+        auto chunk_size_buffer = co_await stream.read_exactly(sizeof(size_t));
+        if (chunk_size_buffer.empty()) {
+            break;
+        }
+
+        iobuf c;
+        c.append(std::move(chunk_size_buffer));
+        auto size_parser = iobuf_parser(std::move(c));
+
+        auto chunk_size = serde::read<size_t>(size_parser);
+        iobuf chunk;
+        chunk.append(co_await stream.read_exactly(chunk_size));
+
+        auto chunk_parser = iobuf_parser{std::move(chunk)};
+        auto hashes = serde::read<fragmented_vector<uint64_t>>(chunk_parser);
+        vlog(_ctxlog.trace, "read {} path hashes from disk", hashes.size());
+
+        for (auto hash : hashes) {
+            if (unlikely(_path_hashes.contains(hash))) {
+                _possible_collisions.insert(hash);
+            } else {
+                _path_hashes.insert(hash);
```

Comment:
nit: could use `[iter, inserted] = hashes.insert(hash)`, and insert to possible_collisions if `inserted=false` so we only do the hash lookup once in the happy path?